core/lib/mempool/src/mempool_store.rs

```diff
@@ -221,22 +221,53 @@ impl MempoolStore {
     }
 
     fn gc(&mut self) -> Vec<Address> {
-        if self.size >= self.capacity {
-            let index: HashSet<_> = self
+        if self.size > self.capacity {
+            let mut transactions = std::mem::take(&mut self.l2_transactions_per_account);
+            let mut possibly_kept: Vec<_> = self
                 .l2_priority_queue
                 .iter()
-                .map(|pointer| pointer.account)
+                .rev()
+                .filter_map(|pointer| {
+                    transactions
+                        .remove(&pointer.account)
+                        .map(|txs| (pointer.account, txs))
+                })
                 .collect();
-            let transactions = std::mem::take(&mut self.l2_transactions_per_account);
-            let (kept, drained) = transactions
+            let mut number_of_accounts_kept = possibly_kept
+                .iter()
+                .scan(0, |sum, (_, txs)| {
+                    *sum += txs.len();
+                    (*sum <= self.capacity as usize).then_some(())
+                })
+                .count();
```

Comment:
Bikeshedding: IMO, `scan` is not the most obvious choice here; an imperative approach with a `for` loop would be better. Alternatively, maybe [`take_while`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take_while) would work here?

Reply:
pushed imperative approach